### PR TITLE
Fix pruning

### PR DESCRIPTION
**Pull request by varunagrawal**
Holding PR for getting the decision tree pruning to work.

### CODE REVIEW

**Comment by varunagrawal** [comment]
@dellaert @ProfFan please review this PR. It should be easier to do so since the intermediate changes (like comments, extraneous prints etc) are all removed.

**Comment by dellaert** [comment]
> @dellaert @ProfFan please review this PR. It should be easier to do so since the intermediate changes (like comments, extraneous prints etc) are all removed.

I will once you remove changes in discrete. That has been my comment over and over again and every time I get back to the review it's still there.

File: gtsam/hybrid/IncrementalHybrid.cpp
Comment (varunagrawal) [inline gtsam/hybrid/IncrementalHybrid.cpp]: Explain what is happening in `prune`. Also, make functional.

File: gtsam/hybrid/IncrementalHybrid.cpp
Comment (varunagrawal) [inline gtsam/hybrid/IncrementalHybrid.cpp]: Explain when does this happen better.

File: gtsam/hybrid/IncrementalHybrid.cpp
Comment (varunagrawal) [inline gtsam/hybrid/IncrementalHybrid.cpp]: // pruneBayesNet sets the leaves with 0 in `discreteFactor` to nullptr in all the conditionals with the same keys in `bayesNetFragment`.

File: gtsam/hybrid/tests/testIncrementalHybrid.cpp
Comment (varunagrawal) [inline gtsam/hybrid/tests/testIncrementalHybrid.cpp:422]: // The result is a HybridBayesNet with no discrete variables (equivalent to a GaussianBayesNet).

Comment (varunagrawal) [reply gtsam/hybrid/tests/testIncrementalHybrid.cpp:422]: Factorization is
`P(... | measurements) = P(W0|Z0) P(Z0|Y0) P(Y0|X0) P(X0)`

File: gtsam/hybrid/tests/testIncrementalHybrid.cpp
Comment (varunagrawal) [inline gtsam/hybrid/tests/testIncrementalHybrid.cpp:484]: // The result is a HybridBayesNet with 1 discrete variable M(1).
// `P(... | measurements) = P(W0|Z0, W1, M1) P(Z0|Y0, W1, M1) P(Y0|X0, W1, M1) P(X0 | X1, W1, M1) P(W1|Z1, X1, M1)P(Z1|Y1, X1, M1)P(Y1 | X1, M1)P(X1 | M1)P(M1)`

Comment (varunagrawal) [reply gtsam/hybrid/tests/testIncrementalHybrid.cpp:484]: TODO(Frank): Fix printing of `GaussianConditional` to show the conditioned variables.

Comment (varunagrawal) [reply gtsam/hybrid/tests/testIncrementalHybrid.cpp:484]: What is the MHS tree here?

File: gtsam/hybrid/tests/testIncrementalHybrid.cpp
Comment (varunagrawal) [inline gtsam/hybrid/tests/testIncrementalHybrid.cpp:535]: What is the factorization?

// `P(... | measurements) = P(W0|Z0, W1, M1) P(Z0|Y0, W1, M1) P(Y0|X0, W1, M1) P(X0 | X1, W1, M1) P(W1|W2, Z1, X1, M1, M2) P(Z1| W2, Y1, X1, M1, M2)P(Y1 | W2, X1, M1, M2)P(X1 | W2, X2, M1, M2) P(W2|Z2, X2, M1, M2) P(Z2|Y2, X2, M1, M2)P(Y2 | X2, M1, M2)P(X2 | M1, M2)  P(M1, M2)`

File: gtsam/hybrid/IncrementalHybrid.cpp
Comment (varunagrawal) [inline gtsam/hybrid/IncrementalHybrid.cpp]: TODO(Frank) This should be a const method on HybridBayesNet.

File: gtsam/hybrid/tests/testIncrementalHybrid.cpp
Comment (varunagrawal) [reply gtsam/hybrid/tests/testIncrementalHybrid.cpp:535]: The MHS at this point should be a 3 level tree on (0, 1, 2).
0 has 2 choices, 1 has 4 choices and 2 has 4 choices.

**Review by dellaert — APPROVED** [review]
Very cool!

**Comment by dellaert** [comment]
But please merge develop into both branches before merging and see if there are conflicts?

**Review by dellaert — CHANGES_REQUESTED** [review]
Same issue with this PR: many unrelated changes, seems you merged in develop in only one of the brancches. Please fix and I'll review.

File: gtsam/hybrid/HybridBayesNet.h
Comment (dellaert) [inline gtsam/hybrid/HybridBayesNet.h]: Why do we have non-const method here? Please re-evaluate.

File: gtsam/discrete/DiscreteKey.h
Comment (dellaert) [inline gtsam/discrete/DiscreteKey.h]: prefer not to have changes outside hybrid. Make static function in cpp file where you need it?

File: gtsam/hybrid/IncrementalHybrid.h
Comment (dellaert) [inline gtsam/hybrid/IncrementalHybrid.h]: not const. I assume this will be resolved in #44

**Review by dellaert — APPROVED** [review]
Approve but only if *both* non-const issues are resolved in #44

File: gtsam/hybrid/IncrementalHybrid.h
Comment (varunagrawal) [reply gtsam/hybrid/IncrementalHybrid.h]: Resolved

File: gtsam/hybrid/HybridBayesNet.h
Comment (varunagrawal) [reply gtsam/hybrid/HybridBayesNet.h]: Removed.

File: gtsam/discrete/DiscreteKey.h
Comment (dellaert) [reply gtsam/discrete/DiscreteKey.h]: prefer not to have changes outside hybrid. Make static function in cpp file where you need it?

**Review by dellaert — CHANGES_REQUESTED** [review]

**Review by dellaert — APPROVED** [review]
Cool! Can't wait to see real results :-)

**Pull request merged**
